wandb/analytics/sentry.py

```diff
@@ -1,213 +1,216 @@
 
-    @functools.wraps(func)
-    def wrapper(self: type[Sentry], *args: Any, **kwargs: Any) -> Any:
-        if self._disabled:
+        # If this instance belongs to a different process (fork happened),
+        # do nothing; singleton() will create a fresh instance for the child.
+        if self._pid != os.getpid():
             return None
+
+        if not self._booted:
+            self._boot()
+
         try:
-            return func(self, *args, **kwargs)
+            return method(self, *args, **kwargs)
         except Exception as e:
-            # do not call self.exception here to avoid infinite recursion
-            if func.__name__ != "exception":
-                self.exception(f"Error in {func.__name__}: {e}")
+            if method.__name__ != "exception":
+                # Best-effort logging of wrapper-level failures.
+                try:
+                    self.exception(f"Error in {method.__name__}: {e}")
+                except Exception:
+                    pass
             return None
 
     return wrapper
 
 
 class Sentry:
-    _disabled: bool
-
-    def __init__(self) -> None:
-        self._disabled = not wandb.env.error_reporting_enabled()
-        self._sent_messages: set = set()
+    def __init__(self, *, pid: int) -> None:
+        from wandb import env as _env
 
-        self.dsn = os.environ.get(wandb.env.SENTRY_DSN, SENTRY_DEFAULT_DSN)
+        self._pid: int = pid
+        self._enabled: bool = bool(_env.error_reporting_enabled())
+        self._booted: bool = False
+        self._boot_lock = threading.Lock()
+        self._atexit_registered: bool = False
 
-        self.scope: sentry_sdk.scope.Scope | None = None
+        self._sent_messages: set[str] = set()
+        self._sdk: Any | None = None  # will hold the sentry_sdk module after boot
+        self.scope: Any | None = None
 
-        # ensure we always end the Sentry session
-        atexit.register(self.end_session)
+        self.dsn: str | None = os.environ.get(_env.SENTRY_DSN, SENTRY_DEFAULT_DSN)
 
     @property
     def environment(self) -> str:
-        """Return the environment we're running in."""
-        # check if we're in a git repo
         is_git = pathlib.Path(__file__).parent.parent.parent.joinpath(".git").exists()
-
-        # these match the environments for gorilla
         return "development" if is_git else "production"
 
-    @_safe_noop
-    def setup(self) -> None:
-        """Setup Sentry SDK.
-
-        We use lower-level APIs (i.e., not sentry_sdk.init) here
-        to avoid the possibility of interfering with the user's
-        own Sentry SDK setup.
-        """
-        client = sentry_sdk.Client(
-            dsn=self.dsn,
-            default_integrations=False,
-            environment=self.environment,
-            release=wandb.__version__,
-        )
-        self.scope = sentry_sdk.get_global_scope().fork()
-        self.scope.clear()
-        self.scope.set_client(client)
+    def _boot(self):
```

Comment:
```suggestion
    def _boot(self) -> None:
```

or else this isn't typechecked